### PR TITLE
Add circuit.tensor

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Add the ``QuantumCircuit.tensor`` method, allowing to tensor another circuit with an
existing circuit. This method works analogously to `Operator.tensor`.

With this method we can deprecate ``QuantumCircuit.extend`` and ``QuantumCircuit.combine`` as stacking/tensoring circuits is still possible.

### Details and comments

For example

```python
>>> from qiskit import QuantumCircuit
>>> top = QuantumCircuit(1)
>>> top.x(0);
>>> bottom = QuantumCircuit(2)
>>> bottom.cry(0.2, 0, 1);
>>> bottom.tensor(top).draw()
        ┌───┐
q_0: ───┤ X ├───
        └───┘
q_1: ─────■─────
     ┌────┴────┐
q_2: ┤ RY(0.2) ├
     └─────────┘
```

### CODE REVIEW

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -687,6 +687,46 @@ def compose(self, other, qubits=None, clbits=None, front=False, inplace=False):
...
 
+    def tensor(self, other):
+        """Tensor ``self`` with ``other``.
+
+        Args:
+            other (QuantumCircuit): The other circuit to tensor this circuit with.
+
+        Returns:
+            QuantumCircuit: the tensored circuit (returns None if inplace==True).
+        """
```
Comment (kdk) [inline qiskit/circuit/quantumcircuit.py:698]: Can you add a note/example to the docstring on the endianness convention? What you have in the release note would be good.

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -687,6 +687,46 @@ def compose(self, other, qubits=None, clbits=None, front=False, inplace=False):
...
+            dest = QuantumCircuit(*other.qregs, *self.qregs, cr)
+
+        # Now we don't have to handle any more cases arising from special implicit naming
+        else:
+            dest = QuantumCircuit(*other.qregs, *self.qregs, *other.cregs, *self.cregs)
+
+        # compose self onto the output, and then other
+        dest.compose(other, range(other.num_qubits), range(other.num_clbits), inplace=True)
+        return dest.compose(self, range(other.num_qubits, num_qubits),
+                            range(other.num_clbits, num_clbits))
```
Comment (kdk) [inline qiskit/circuit/quantumcircuit.py:728]:
```suggestion
        dest.compose(self, range(other.num_qubits, num_qubits),
                            range(other.num_clbits, num_clbits), inplace=True)
        return dest
```
Can the second compose also be `inplace=True`? (It'd save an unneeded copy of `dest`.)

**Review by kdk — COMMENTED** [review]
LGTM, two small comments.

File: qiskit/circuit/quantumcircuit.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:727]: Added in 340f673 👍

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -690,6 +690,76 @@ def compose(self, other, qubits=None, clbits=None, front=False, inplace=False):
 
         return dest
 
+    def tensor(self, other):
+        """Tensor ``self`` with ``other``.
+
+        Remember that in the little-endian convention the leftmost operation will be at the bottom
+        of the circuit. See also
+        [the docs](qiskit.org/documentation/tutorials/circuits/1_getting_started_with_qiskit.html)
```
Comment (kdk) [inline qiskit/circuit/quantumcircuit.py:698]:
```suggestion
        [the docs](qiskit.org/documentation/tutorials/circuits/3_summary_of_quantum_operations.html)
```

**Review by kdk — APPROVED** [review]

**Pull request merged**